packages/react-kit/src/lib/signer/useCallSignerFromIframe.ts

```diff
@@ -0,0 +1,122 @@
+      if (event.origin === childIframeOrigin) {
+        const functionName = event.data.function;
+        if (functionName === "wait") {
+          const commonPostMessage = {
+            function: functionName,
+            originalFunction: event.data.originalFunction
+          };
+          try {
+            const txResponseWithoutWait = event.data.txResponse;
+            const args = event.data.args ?? [];
+            const txResponseWithWait = txMap.get(txResponseWithoutWait.hash);
+            if (txResponseWithWait) {
+              const txReceipt = await txResponseWithWait.wait(...args);
+              iframeRef.current?.contentWindow?.postMessage(
+                {
+                  ...commonPostMessage,
+                  result: txReceipt
+                },
+                "*"
+              );
+            }
+          } catch (error) {
+            console.error(error);
+            const eventMessage = {
+              ...commonPostMessage,
+              error:
+                error && typeof error === "object" && "message" in error
+                  ? error.message
+                  : error
+            };
+            iframeRef.current?.contentWindow?.postMessage(eventMessage, "*");
+          }
+        } else if (functionName) {
+          const fn: keyof typeof EthersAdapter["prototype"] = functionName;
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const args: any[] = event.data.args ?? [];
+
+          try {
+            const result = await ethersSigner?.[fn]?.(args[0], args[1]);
+            if (
+              functionName === "sendTransaction" &&
+              typeof result === "object" &&
+              "wait" in result
+            ) {
+              const txResponse = result;
+              // the wait function can not be serialized when sending the txResponse over postMessage so we exclude it
+              // eslint-disable-next-line @typescript-eslint/no-unused-vars
+              const { wait, ...txResponseWithoutWait } = txResponse;
+              txMap.set(txResponse.hash, txResponse);
+              const eventMessage = {
+                function: functionName,
+                result: txResponseWithoutWait
+              };
+              iframeRef.current?.contentWindow?.postMessage(eventMessage, "*");
+            } else {
+              const eventMessage = {
+                function: functionName,
+                result
+              };
+              iframeRef.current?.contentWindow?.postMessage(eventMessage, "*");
+            }
+          } catch (error) {
+            console.error(error);
+            const eventMessage = {
+              function: functionName,
+              error:
+                error && typeof error === "object" && "message" in error
+                  ? error.message
+                  : error
+            };
+            iframeRef.current?.contentWindow?.postMessage(eventMessage, "*");
```

Comment:
Everywhere we use `postMessage(..., "*")` I think we should find a way to avoid posting the message to every window.
A message posted by the widget to the hosting webpage could use parentOrigin
A response posted by the parentOrigin to the widget could use the widget host

Reply:
fixed